src/rust/engine/src/externs/interface.rs

```diff
@@ -1647,6 +1647,7 @@ fn write_digest(
   digest: &PyAny,
   path_prefix: String,
   clear_paths: Vec<String>,
```

Comment:
Should this also invalidate these deleted paths? If so, this will often be a directory: does invalidating a (deleted) parent directory invalidate any children?

This was added in #18930 for goals like `package` and `export-codegen` which don't write to files in the graph (AIUI), but potentially can be used for writing to tracked files too with something like https://github.com/pantsbuild/pants/discussions/18235

Reply:
> If so, this will often be a directory: does invalidating a (deleted) parent directory invalidate any children?

Deleted directories should be invalidated, since the `scandir` of their parent needs to be updated. But no, invalidation is not recursive: it's for exact match paths (since all of our filesystem operations are syscalls on individual paths.

Reply:
Ah, so if this does a recursive delete, I need to extract each path and invalidate them individually. Okay, thanks.

Reply:
Ah, that's potentially unhelpful for performance, e.g. `pants package :loose-pex` will clear `dist/loose-pex.pex/` recursively, and that may contain thousands of little files (each unpacked wheel), which would be fine if we _have_ to do it... but seems a bit unfortunate for paths like `dist/...` that are aren't in the graph.

That said, I _imagine_ the cost of actually doing the deletes is far greater than the time to do a no-op bulk invalidation of many paths, so I'll ignore it for now.

Reply:
There is no need to recursively invalidate below a deleted directory, because all `Snapshots` for a subdirectory depend on the `scandir` of all parent nodes (in order to determine whether any of the path components were symlinks).